File: src/main/java/no/novari/configuration/model/configuration/entities/collection/ObjectsFromCollectionMapping.java
package no.novari.configuration.model.configuration.entities.collection;

import com.fasterxml.jackson.annotation.JsonIgnore;
import jakarta.persistence.CascadeType;
import jakarta.persistence.Entity;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.JoinTable;
import jakarta.persistence.OneToMany;
import jakarta.persistence.OneToOne;
import jakarta.validation.Valid;
import jakarta.validation.constraints.NotEmpty;
import jakarta.validation.constraints.NotNull;
import lombok.AccessLevel;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import lombok.extern.jackson.Jacksonized;
import no.novari.configuration.model.configuration.entities.ObjectMapping;

import java.util.Collection;

@Getter
@Builder
@Jacksonized
@AllArgsConstructor
@NoArgsConstructor
@Entity
public class ObjectsFromCollectionMapping {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @JsonIgnore
    @Setter(AccessLevel.NONE)
    private long id;

    @OneToMany(cascade = CascadeType.ALL, orphanRemoval = true)
    @JoinTable(
            name = "objects_from_collection_mapping_references_ordered",
            joinColumns = @JoinColumn(name = "objects_from_collection_mapping_id"),
            inverseJoinColumns = @JoinColumn(name = "instance_collection_reference_id")
    )
    @NotEmpty
    private Collection<@Valid @NotNull InstanceCollectionReference> instanceCollectionReferencesOrdered;

    @OneToOne(cascade = CascadeType.ALL, orphanRemoval = true)
    @Valid
    @NotNull
    private ObjectMapping objectMapping;

}

File: src/main/kotlin/no/novari/flyt/configuration/model/configuration/entities/collection/ObjectsFromCollectionMapping.kt
package no.novari.flyt.configuration.model.configuration.entities.collection

import com.fasterxml.jackson.annotation.JsonIgnore
import jakarta.persistence.CascadeType
import jakarta.persistence.Entity
import jakarta.persistence.GeneratedValue
import jakarta.persistence.GenerationType
import jakarta.persistence.Id
import jakarta.persistence.JoinColumn
import jakarta.persistence.JoinTable
import jakarta.persistence.OneToMany
import jakarta.persistence.OneToOne
import jakarta.validation.Valid
import jakarta.validation.constraints.NotEmpty
import jakarta.validation.constraints.NotNull
import no.novari.flyt.configuration.model.configuration.entities.ObjectMapping

@Entity
class ObjectsFromCollectionMapping(
    @field:Id
    @field:GeneratedValue(strategy = GenerationType.IDENTITY)
    @field:JsonIgnore
    var id: Long = 0,
    @field:OneToMany(cascade = [CascadeType.ALL], orphanRemoval = true)
    @field:JoinTable(
        name = "objects_from_collection_mapping_references_ordered",
        joinColumns = [JoinColumn(name = "objects_from_collection_mapping_id")],
        inverseJoinColumns = [JoinColumn(name = "instance_collection_reference_id")],
    )
    @field:NotEmpty
    var instanceCollectionReferencesOrdered: MutableCollection<
        @Valid @NotNull
        InstanceCollectionReference,
    > =
        mutableListOf(),
    @field:OneToOne(cascade = [CascadeType.ALL], orphanRemoval = true)
    @field:Valid
    @field:NotNull
    var objectMapping: ObjectMapping? = null,
) {
    companion object {
        @JvmStatic
        fun builder(): Builder = Builder()
    }

    class Builder {
        private var id: Long = 0
        private var instanceCollectionReferencesOrdered: MutableCollection<InstanceCollectionReference> =
            mutableListOf()
        private var objectMapping: ObjectMapping? = null

        fun id(id: Long) = apply { this.id = id }

        fun instanceCollectionReferencesOrdered(
            instanceCollectionReferencesOrdered: Collection<InstanceCollectionReference>?,
        ) = apply {
            this.instanceCollectionReferencesOrdered =
                instanceCollectionReferencesOrdered?.toMutableList() ?: mutableListOf()
        }

        fun objectMapping(objectMapping: ObjectMapping?) = apply { this.objectMapping = objectMapping }

        fun build(): ObjectsFromCollectionMapping =
            ObjectsFromCollectionMapping(
                id = id,
                instanceCollectionReferencesOrdered = instanceCollectionReferencesOrdered,
                objectMapping = objectMapping,
            )
    }
}
